File: Solutions/189_Rotate_Array/189_Rotate_Array.cpp
class Solution {
public:
void rotate(vector<int>& nums, int k) {
	k = k % nums.size();
	vector<int> temp(k);
	
	int j = 0;
	// Store the last k elements in the temporary array
	for (int i=nums.size()-k; i<nums.size(); i++) {
		temp[j++] = nums[i];
	}

	j = nums.size() - 1;
	// Right shift all the elements by k steps
	for (int i=nums.size()-k-1; i>=0; i--) {
		nums[j--] = nums[i];
	}
	
	// Populate the first k elements with the temporary array elements
	for (int i=0; i<k; i++) {
		nums[i] = temp[i];
	}
}
};

//Second fastest
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
       
        int n=nums.size(),j=0;
         k=k%n;
        vector<int>r;
        for(int i=n-k;i<n;i++)
        {
            r.push_back(nums[i]);
            
        }
          for(int i=0;i<n-k;i++)
        {
            r.push_back(nums[i]);
        }
        for(int i=0;i<n;i++)
        {
            nums[i]=r[i];
        }
    
    }
};

//Fastest
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        k = k % nums.size();
        int count = 0;
        for (int start = 0; count < nums.size(); start++) {
          int current = start;
          int prev = nums[start];
          do {
            int next = (current + k) % nums.size();
            int temp = nums[next];
            nums[next] = prev;
            prev = temp;
            current = next;
            count++;
          } while (start != current);
        }
    }
};

File: Solutions/189_Rotate_Array/189_Rotate_Array.go
//Second fastest
func rotate(nums []int, k int)  {
    var (
        l = len(nums)
    )
    if k >= l {
        k = k % l
    }
    if k == 0 {
        return
    }
    
    var (
        tmp int
        dem int
        i int
        j int
        v int
    )
    
    // move element to the right if k is less than half of the number of elements
    if k <= l/2 {
        i = k
        v = nums[j]
        for dem < l {
            tmp = nums[i]
            nums[i] = v
            dem++
            if i == j {
                j++
                v = nums[j]
                i = k+j
                continue
            }
            i = (i+k) % l
            v = tmp
        }
        return
    }
    
    // move elements to the left
    k = l-k
    i = 0
    j = k
    v = nums[j]
    for dem < l {
        tmp = nums[i]
        nums[i] = v
        dem++
        if i == j {
            j++
            v = nums[j]
            i = (j-k+l)% l
            continue
        }
        i = (i-k+l)% l
        v = tmp
    }
}

//Fastest
func rotate(nums []int, k int)  {
    length := len(nums)
	if length < 2 {
		return
	}

	k = k % length

	reverse(nums,  length - k, length - 1)
	reverse(nums, 0, length - k - 1)
	reverse(nums,  0, length - 1)
}

func reverse(nums[] int, start int, end int) {
	count := (end - start + 1) / 2
	for i := 0; i < count; i++ {
		current := i + start
		j := end - i
		nums[current] += nums[j]
		nums[j] = nums[current] - nums[j]
		nums[current] -= nums[j]
	}
}